Set up Vault and use Vault-generated certs

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/747
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This sets up vault, and then sets consul to use vault-generated certificates.

This currently works without policies since we're using -dev for consul and vault. For prod we'll need to set up policies first.

Unfortunately there's some extra junk around polling/sleeping since this change does make it longer for all the services to boot up
<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
e2e and integration tests. make up also works

<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Run Vault as part of local grapl

This starts up Vault upon make up and removes it upon make down. Since
we use Vault in AWS, this is a step towards making our dev environments
closer to what prod will look like

Commit:
Set up conf file for Vault

Commit:
Fix the vault config in dev

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Error check for vault failures

Commit:
WIP: Fixes to vault

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
WIP wait for Vault to boot prior to configuring

Commit:
Poll for consul to be ready

When using vault-generated certs, consul takes a bit of time to be
ready for services. If its not ready, nomad gets an odd constraint
error where it can't seem to read the consul version properly. The
right way to deal with this is to poll until consul is ready. We're
using the number of services as a quick n dirty way to check for
readiness. There should be a better way, so this will probably be
revisited in the future

twunderlich-grapl commented on nomad/local/vault-agent-conf.hcl:
```diff
@@ -0,0 +1,18 @@
+
+api_addr = "http://127.0.0.1:8200"
+
+#service_registration "consul" {
```
ugh, so there's a circular dependency here. We need vault to configure consul, but that means we can't use the service_registration stanza here since that expects consul to be ready immediately 💩 
I'm looking for cli commands to do this now

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Remove vault service_registration stanza

wimax-grapl commented on nomad/local/start_detach.sh:
Is there a more robust way to tell if Vault is ready?

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Merge branch 'main' into twunderlich/vault-integration

Commit:
Lower the sleep amount

I spent some time trying to find the right thing to poll on. I was
successful in polling for vault to be ready, but unfortunately the
main issue is that there's a race condition where Nomad can't get the
consul version. Nomad is able to actually assign the allocs but nomad
up exits due to the previous error even though Nomad can recover. Oh
well.

Commit:
Collect vault logs

Commit:
Poll for vault being ready

Commit:
Debug info

Commit:
test using new token location

Commit:
Start consul prior to Nomad to avoid race condition

We were hitting a race condition where Nomad errored because it
couldn't get the consul version. This is fixed by starting consul
prior to starting Nomad. This is actually how hashicorp does it in
their systemd scripts

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Add debug info

Commit:
Rm silent

Commit:
More debug output

Commit:
Merge branch 'main' into twunderlich/vault-integration

 Conflicts:
	nomad/local/start_detach.sh

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Add vault status error check

Commit:
Clear out old logs

Commit:
Move vault token up

Commit:
Remove comments

twunderlich-grapl commented on nomad/local/start_detach.sh:
```diff
@@ -65,24 +74,82 @@ create_dynamic_consul_config() {
 EOF
 }
 
+clear_hashicorp_log_files() {
```
Guess who was trying to debug based on logs from some other build? 😠

Commit:
Don't error out if the logs are missing

Commit:
Add debug statements

Commit:
Comment out sleep

Commit:
Back out debug statements

Commit:
Merge remote-tracking branch 'origin/main' into twunderlich/vault-integration

Commit:
Refactor to be cleaner

Commit:
Poll correctly

Commit:
Check the exit code

Commit:
Lose the pipefail

Commit:
Ugh why do you exit like this

Commit:
Anger leads to hate.

Commit:
Vault status LIES

Commit:
refactor clearing log files

Commit:
Cleanup work

twunderlich-grapl commented on etc/ci_scripts/dump_artifacts/nomad_artifacts.py:
```diff
@@ -53,9 +54,15 @@ def dump_all(artifacts_dir: Path, dump_agent_logs: bool) -> None:
+    ]
+    for log_path in hashicorp_log_paths:
+        if os.path.exists(log_path):
+            shutil.copy2(log_path, artifacts_dir)
```
If something fails prior to nomad starting, you better believe I want the logs

Commit:
Do I really need to sleep?

Commit:
Simplify loop logic

Since we needed the set +x, we can simply just check if the command
failed

twunderlich-grapl commented on nomad/local/start_detach.sh:
```diff
@@ -65,24 +74,83 @@ create_dynamic_consul_config() {
+                # vault status returns an exit code of 0 for unsealed (ie ready), 1 for error and 2 for sealed
+                # Since we only want to capture the exit code, we need to drop all output from the command
+                vault_exit_code=\$(vault status &>/dev/null; echo $?)
+                while [[  \${vault_exit_code} != 0 ]]; do
```
Thanks! I was doing all that extra work to avoid the script exiting, but after running set +x that didn't matter. Fixed
